hal.zig

```diff
@@ -8,11 +8,91 @@ pub const Pin = enum(u8) { // or whatever fits for the HAL
 
     /// Runtime lookup of a pin by name.
     pub fn resolve(name: []const u8) ?Pin;
+
+    /// The corresponding UART instance this pin can be configured to.
+    /// - However... What happens when multiple Uarts can be routed to the same pin?
+    /// - There's an argument this function shouldn't exist, and it's on the user
+    ///   to correctly configure the corresponding pins before using the UART driver,
+    ///   and specify the correct Uart instance themselves.
+    pub fn to_uart_instance(self: Pin) !uart.Instance;
```

Comment:
Imho, "Pin to peripherial conversion" is the wrong way, and we should either go the other way round or just provide a generic purpose function selector or something like that